src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java

```diff
@@ -366,31 +398,47 @@ public ObjectId getLastCommitHash(URL repoUrl) throws EntityNotFoundException {
      * Stager Task #3: Filter late submissions Filter all commits after exercise due date
      *
      * @param repository Local Repository Object.
-     * @param exercise   ProgrammingExercise associated with this repo.
+     * @param participation   ProgrammingExerciseStudentParticipation associated with this repo.
+     * @param filterLateSubmissionsDate the date after which all submissions should be filtered out (may be null)
      */
-    public void filterLateSubmissions(Repository repository, ProgrammingExercise exercise) {
-        if (exercise.getDueDate() == null) {
-            // No dates set on exercise
+    @Transactional(readOnly = true)
+    public void filterLateSubmissions(Repository repository, ProgrammingExerciseStudentParticipation participation, ZonedDateTime filterLateSubmissionsDate) {
+        if (filterLateSubmissionsDate == null) {
+            // No date set in client and exercise has no due date
             return;
         }
 
         try {
             Git git = new Git(repository);
+            Optional<Submission> lastValidSubmission = participation.getSubmissions().stream().filter(s -> s.getSubmissionDate().isBefore(filterLateSubmissionsDate))
```

Comment:
```suggestion
            Optional<Submission> lastValidSubmission = participation.getSubmissions().stream().filter(s -> s.getSubmissionDate().isBefore(filterLateSubmissionsDate)).max(Comparator.comparing(Submission::getSubmissionDate));
```

This is a recommendation from Intellij.